app/upgrades/v15/upgrades_test.go

```diff
@@ -0,0 +1,71 @@
+package v15_test
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	tmproto "github.com/cometbft/cometbft/proto/tendermint/types"
+
+	"github.com/cosmos/cosmos-sdk/testutil/mock"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	slashingtypes "github.com/cosmos/cosmos-sdk/x/slashing/types"
+
+	"github.com/cosmos/gaia/v15/app/helpers"
+	v15 "github.com/cosmos/gaia/v15/app/upgrades/v15"
+)
+
+func TestUpgradeSigningInfos(t *testing.T) {
+	gaiaApp := helpers.Setup(t)
+	ctx := gaiaApp.NewUncachedContext(true, tmproto.Header{})
+	slashingKeeper := gaiaApp.SlashingKeeper
+
+	signingInfosNum := 8
+	emptyAddrCtr := 0
+
+	// create some dummy signing infos, half of which with an empty address field
+	for i := 0; i < signingInfosNum; i++ {
+		pubKey, err := mock.NewPV().GetPubKey()
+		require.NoError(t, err)
+
+		consAddr := sdk.ConsAddress(pubKey.Address())
+		info := slashingtypes.NewValidatorSigningInfo(
+			consAddr,
+			0,
+			0,
+			time.Unix(0, 0),
+			false,
+			0,
+		)
+
+		if i <= signingInfosNum/2 {
+			info.Address = ""
+			emptyAddrCtr++
+		}
+
+		slashingKeeper.SetValidatorSigningInfo(ctx, consAddr, info)
+		require.NoError(t, err)
+	}
+
+	// check signing info were correctly created
+	slashingKeeper.IterateValidatorSigningInfos(ctx, func(address sdk.ConsAddress, info slashingtypes.ValidatorSigningInfo) (stop bool) {
```

Comment:
I believe that the test correctly checks for the upgrade results but it is difficult to follow.

That being said, I cannot se the before and after cases. I would be much more confident if I the test explicitly compared signing infos before and after the upgrade. Relying on `require.Zero(t, emptyAddrCtr)` can work, it's just difficult to figure out.

Would it be possible to refactor this in such a way as to keep the "before" signing infos slice and compare it with "after" signing infos slice?

Reply:
Thanks for the feedback.  Test simplified in [7e58f6b](https://github.com/cosmos/gaia/pull/2886/commits/7e58f6b5de6cc0a02351043027007b87c96817d6).

Reply:
Thanks!